python/ray/dashboard/subprocesses/handle.py

```diff
@@ -204,206 +190,121 @@ async def _do_periodic_health_check(self):
+        Proxy handler for WebSocket API
+        It establishes a WebSocket connection with the client and simultaneously connects
+        to the backend server's WebSocket endpoint. Messages are forwarded in both directions.
         """
-        try:
-            response = await asyncio.wrap_future(prev_fut)
-            await response.write(str(exception).encode())
-            await response.write_eof()
-            response_fut.set_result(response)
-        except Exception as e:
-            response_fut.set_exception(e)
-
-    def handle_parent_bound_message(self, message: ParentBoundMessage):
-        """Handles a message from the parent bound queue. This function must run on a
-        dedicated thread, called by dispatch_parent_bound_messages."""
-        loop = self.loop
-        if isinstance(message, UnaryResponseMessage):
-            active_request = self.active_requests.pop_or_raise(message.request_id)
-            # set_result is not thread safe.
-            loop.call_soon_threadsafe(
-                active_request.response_fut.set_result,
-                aiohttp.web.Response(
-                    status=message.status,
-                    body=message.body,
-                ),
-            )
-        elif isinstance(message, StreamResponseStartMessage):
-            active_request = self.active_requests.get_or_raise(message.request_id)
-            assert active_request.stream_response is None
-            # This assignment is thread safe, because a next read will come from another
-            # handle_parent_bound_message call for a Stream.*Message, which will run on
-            # the same thread and hence will happen-after this assignment.
-            active_request.stream_response = asyncio.run_coroutine_threadsafe(
-                SubprocessModuleHandle.handle_stream_response_start(
-                    active_request.request, message.body
-                ),
-                loop,
+        ws_from_client = aiohttp.web.WebSocketResponse()
+        await ws_from_client.prepare(request)
+
+        url = f"http://localhost{request.path_qs}"
+
+        async with self.session.ws_connect(
+            url, headers=filter_headers(request.headers)
+        ) as ws_to_backend:
+
+            async def forward(ws_source, ws_target):
+                async for msg in ws_source:
+                    if msg.type == aiohttp.WSMsgType.TEXT:
+                        await ws_target.send_str(msg.data)
+                    elif msg.type == aiohttp.WSMsgType.BINARY:
+                        await ws_target.send_bytes(msg.data)
+                    else:
+                        logger.error(f"Unknown msg type: {msg.type}")
+                        await ws_target.close()
+
+            # Set up two tasks to forward messages in both directions.
+            task_client_to_backend = self.loop.create_task(
+                forward(ws_from_client, ws_to_backend)
             )
```

Comment:
why do we need this direction?

Reply:
Websocket is bi-directional. I don't know whether we want to support this direction in the future although it is not used for now.